sdk/identity/azure-identity/azure/identity/aio/_credentials/token_exchange.py

```diff
@@ -0,0 +1,43 @@
+# ------------------------------------
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+# ------------------------------------
+import os
+from typing import TYPE_CHECKING
+
+from .client_assertion import ClientAssertionCredential
+from .._internal import AsyncContextManager
+from ... import CredentialUnavailableError
+from ..._constants import EnvironmentVariables
+from ..._credentials.token_exchange import TokenFileMixin
+
+if TYPE_CHECKING:
+    # pylint:disable=unused-import,ungrouped-imports
+    from typing import Any
+    from azure.core.credentials import AccessToken
+
+
+class TokenExchangeCredential(AsyncContextManager, TokenFileMixin):
+    def __init__(self, **kwargs: "Any") -> None:
+        super().__init__()
+        self._available = all(os.environ.get(var) for var in EnvironmentVariables.TOKEN_EXCHANGE_VARS)
+        if self._available:
+            self._credential = ClientAssertionCredential(
+                os.environ[EnvironmentVariables.AZURE_TENANT_ID],
+                os.environ[EnvironmentVariables.AZURE_CLIENT_ID],
+                self.get_service_account_token,
+                **kwargs
+            )
+
+    async def __aenter__(self):
+        if self._credential:
+            await self._credential.__aenter__()
+        return self
+
+    async def close(self) -> None:
+        await self._credential.close()
```

Comment:
Should we do a check for `self._credential` here (and if we shouldn't, should we do that check in the sync credential)? Also, is there a reason that async credentials follow a `close() -> self._x.close()` pattern instead of `close() -> self.__aexit__() -> self._x.__aexit__()` pattern like sync credentials do?

Reply:
We should check that, thanks for noticing! There's no good reason for the sync vs. async difference. I prefer close -> __exit__ today but it doesn't really matter.